torch/csrc/autograd/python_variable.cpp

```diff
@@ -551,10 +551,14 @@ void initTensorImplConversion(PyObject* module) {
     auto p = c10::intrusive_ptr<c10::TensorImpl, at::UndefinedTensorImpl>::
         unsafe_reclaim_from_nonowning(static_cast<c10::TensorImpl*>(ptr));
     TORCH_CHECK(p.defined(), "Can't wrap undefined tensor");
-    TORCH_CHECK(!p->is_variable(), "Can wrap only non-variable tensor");
     auto tensor = at::Tensor::wrap_tensor_impl(std::move(p));
-    return py::cast(torch::autograd::Variable(
-        torch::autograd::make_variable(std::move(tensor), false)));
+    // For now, there is no guarantee that the tensors returned from Caffe2 ops
```

Comment:
You might be able to avoid this check if you instantiate the outputs in the wrapper correctly.
Currently, the `outputs` variable there is just instantiated to be uninitialized `at::Tensor`s which means they're not variables.

Reply:
@smessmer I think we might still need this check because of the following use case: https://github.com/pytorch/pytorch/blob/10c4b98ade8349d841518d22f19a653a939e260c/caffe2/python/workspace_test.py#L296-L298
Here, the tensor `foo` is created from Caffe2, and it's not a Variable. When `FetchTorch` is run which goes through the `_wrap_tensor_impl` code path, it will find that `foo` is not a Variable, and it will throw error if we directly run `torch::autograd::Variable(tensor)`. Hence I think we still need the check here and do the wrapping when necessary.

Reply:
makes sense, we still need it. It still makes sense to create the preallocated output tensors as variables though, because that's a potential behavior change we want to be sure caffe2 ops can handle.

Reply:
I think it might be a bit tricky to change the `outputs` from `at::Tensor` to `torch::autograd::Variable`, because that would mean that `caffe2/core/export_caffe2_op_to_c10.h` needs to depend on `torch/csrc/autograd/variable.h`, which is not part of the core code.

I think an easier approach might be to wait until Variable/Tensor merge is complete, at which point the `torch::autograd::Variable` class is removed and `at::Tensor` is a Variable automatically (with no autograd metadata), and therefore no change is needed for the `outputs` construction code in `caffe2/core/export_caffe2_op_to_c10.h`.